### ISSUE TITLE
StochasticBlockModel

### ISSUE THREAD

**Issue by dot-at** [JuliaGraphs/LightGraphs.jl#513]
I may be mistaken, but I think the implementation of the **graph generator based on the StochasticBlockModel type is broken**. It looks like you have to give the number of edges to the Graph constructor. The effect is that the **events for the presence of edges are not independent**. However, according to the prevalent understanding of that model (e.g., wikipedia) the edges should be independent.

**Comment by sbromberger** [comment]
The graph generator based on SBM is probably not what you're looking for then. You might want to use the raw `stochastic_block_model()` function. In any case, paging @CarloLucibello :)

**Comment by jpfairbanks** [comment]
This is an implementation difference between the model and implemented code. This implementation represents a compromise between several considerations including performance. If you have an implementation that satisfies your need for independent edges and samples in O(ne) time, I will gladly take a look at it.

**Comment by dot-at** [comment]
It's not an implementation of the stochastic block model. It's an implementation of something else, and that's fine.
Just don't call it StochasticBlockModel if it's something else.

**Comment by sbromberger** [comment]
The `Graph` constructor is not advertised as being a pure stochastic block model constructor. That's what the `stochastic_block_model()` function is designed to be.

I guess I'm not seeing the issue if you're talking about the Graph constructor. Nowhere is it advertised that it's a pure SBM implementation; merely that it uses the SBM as an edge generator. What am I missing?

**Comment by CarloLucibello** [comment]
hi, I missed the notification of last ping.

As Seth said the `stochastic_block_model` method is what you are looking for. The implementation is
O(ne), since the number of edges between blocks is generated in advance sampling from a binomial distribution. This ensures edge independence as well.

I agree that the StochasticBlockModel type could enjoy better docs and maybe a name changing. It should be said though that definitions in some cases are not stable and/or clear cut, scholars sometimes give different names to the same things or the other way around. 

I imagine that a SBM with a fixed number of edges could be admissible. For example, the erdos_renyi model comes into two version, one with independent edges and one with fixed number of edges

**Comment by sbromberger** [comment]
@dot-at - if you'd like to submit a PR to help clarify that the generator / type is not pure SBM, we'd certainly consider it (as @CarloLucibello said, the documentation is imperfect, and as you've demonstrated, it's confusing to the users). Thanks.

**Comment by dot-at** [comment]
Cool, thanks.
I guess the point is that, as a user of the module, I want to trust that something called X is really X. I don't want to compare the fine print of X in the documentation to the Wikipedia definition of X, and realize that it's something else. Does that make sense? It makes using the module difficult.

The binomial and uniform Erdos-Renyi models are only similar if you set the number of edges in the uniform model to the expected number of edges in the binomial model. For SBM, there's no choice of the number of edges which makes the two models similar (unless you consider 2 n^2 log n a reasonable number of edges (cf. coupon collector)).
In summary, it's really cool to have as many different models as possible, and having a fast "approximation" of SBM is certainly great. But the lack of clarity makes the module hard to use.

What's a PR?

**Comment by sbromberger** [comment]
> I guess the point is that, as a user of the module, I want to trust that something called X is really X. I don't want to compare the fine print of X in the documentation to the Wikipedia definition of X....

I'm not sure this is a reasonable expectation across the entire codeset, especially where definitions are frequently in flux and technical decisions favor a specific set of design goals. If this were the case we would need no documentation. It's a nice aspirational goal, though, and therefore one we should TRY to meet. I'm open to ways of improving the documentation to get there.

> What's a PR?

A [pull request](https://help.github.com/articles/about-pull-requests/). Basically, you clone the LightGraphs repo, make changes to your local copy, and then push it up to your github repo, and then there should be a "create pull request" option to propose merging your changes into the main corpus. You then become a contributor once it's merged :)

If creating a PR is too daunting a task, perhaps you could propose the specific changes you'd like to see as a followup comment and we can discuss them here, and then someone can create the PR for you.

**Comment by sbromberger** [comment]
@dot-at just following up here. Were you going to take a shot at a documentation or code change? If not, I'll close this out. Let me know please?

**Comment by sbromberger** [comment]
I'm closing this out in an effort to get the slate clean for v1.0.0 and beyond. @dot-at , let me know if you'd like to discuss further and we can reopen.
